index.js

```diff
@@ -1,216 +1,118 @@
+var nanobus = require('nanobus')
+var assert = require('assert')
 
-  const _store = start._store = barracks()
-  var _router = start._router = null
-  var _routerOpts = null
-  var _rootNode = null
-  var _routes = null
-  var _frame = null
+var onHistoryChange = require('./lib/history')
+var onHref = require('./lib/href')
 
-  if (typeof window !== 'undefined') {
-    _store.use({ onStateChange: render })
-  }
-  _store.use(opts)
-
-  start.toString = toString
-  start.router = router
-  start.model = model
-  start.start = start
-  start.stop = _store.stop
-  start.use = use
-
-  return start
-
-  // render the application to a string
-  // (str, obj) -> str
-  function toString (route, serverState) {
-    serverState = serverState || {}
-    assert.equal(typeof route, 'string', 'choo.app.toString: route must be a string')
-    assert.equal(typeof serverState, 'object', 'choo.app.toString: serverState must be an object')
-    _store.start({ subscriptions: false, reducers: false, effects: false })
-
-    const state = _store.state({ state: serverState })
-    const router = createRouter(_routerOpts, _routes, createSend)
-    const tree = router(route, state)
-    return tree.outerHTML || tree.toString()
-
-    function createSend () {
-      return function send () {
-        assert.ok(false, 'choo: send() cannot be called from Node')
-      }
-    }
-  }
+module.exports = Framework
 
-  // start the application
-  // (str?, obj?) -> DOMNode
-  function start () {
-    _store.model(createLocationModel(opts))
-    const createSend = _store.start(opts)
-    _router = start._router = createRouter(_routerOpts, _routes, createSend)
-    const state = _store.state({state: {}})
+function Framework (opts) {
+  opts = opts || {}
 
-    const tree = _router(state.location.href, state)
-    assert.ok(tree, 'choo.start: the router should always return a valid DOM node')
-    assert.equal(typeof tree, 'object', 'choo.start: the router should always return a valid DOM node')
-    _rootNode = tree
-    tree.done = done
+  var routerOpts = {
+    default: opts.defaultRoute || '/404',
+    curry: true
+  }
 
-    return tree
+  var router = nanorouter(routerOpts)
+  var bus = nanobus()
+  var rerender = null
+  var tree = null
+  var state = {}
 
-    // allow a 'mount' function to return the new node
-    // html -> null
-    function done (newNode) {
-      _rootNode = newNode
-    }
+  return {
+    toString: toString,
+    use: register,
+    mount: mount,
+    route: route,
+    start: start
   }
 
-  // update the DOM after every state mutation
-  // (obj, obj, obj, str, fn) -> null
-  function render (state, data, prev, name, createSend) {
-    if (!_frame) {
-      _frame = nanoraf(function (state, prev) {
-        const newTree = _router(state.location.href, state, prev)
-        _rootNode = yo.update(_rootNode, newTree)
-      })
-    }
-    _frame(state, prev)
+  function route (route, handler) {
+    router.on(route, function (params) {
```

Comment:
How does `toString` then passes the `state` argument around? Alternatively it could overwrite the state I guess.